Sources/NIO/CircularBuffer.swift

```diff
@@ -79,6 +85,80 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
         }
     }
 
+    /// Replaces the specified subrange of elements with the given collection.
+    ///
+    /// - Parameter subrange:
+    /// The subrange of the collection to replace. The bounds of the range must be valid indices of the collection.
+    ///
+    /// - Parameter newElements:
+    /// The new elements to add to the collection.
+    ///
+    /// *O(n)* where _n_ is the length of the new elements collection if the subrange equals to _n_
+    ///
+    /// *O(m)* where _m_ is the combined length of the collection and _newElements_
+    public mutating func replaceSubrange<C>(_ subrange: Range<Index>, with newElements: C) where C : Collection, E == C.Element {
+        precondition(subrange.lowerBound >= self.startIndex && subrange.upperBound <= self.endIndex, "Subrange out of bounds")
+
+        if subrange.count == newElements.count {
+            for (index, element) in zip(stride(from: subrange.lowerBound, to: subrange.upperBound, by: 1), newElements) {
+                self.buffer[self.bufferIndex(ofIndex: index)] = element
+            }
+        } else if subrange.count == self.count && newElements.isEmpty {
+            self.removeSubrange(subrange)
+        } else {
+            var newBuffer: ContiguousArray<E?> = []
+            #if swift(>=4.2)
+            let capacityDelta = (newElements.count - subrange.count)
+            #else
+            let capacityDelta = (Int(newElements.count) - subrange.count)
+            #endif
+            let newCapacity = Int(UInt32(self.buffer.count + capacityDelta).nextPowerOf2())
+            newBuffer.reserveCapacity(newCapacity)
+
+            // This mapping is required due to an inconsistent ability to append sequences of non-optional
+            // to optional sequences.
+            // https://bugs.swift.org/browse/SR-7921
+            newBuffer.append(contentsOf: self[0..<subrange.lowerBound].map { $0 })
+            newBuffer.append(contentsOf: newElements.map { $0 })
+            newBuffer.append(contentsOf: self[subrange.upperBound..<self.endIndex].map { $0 })
+
+            self.tailIdx = newBuffer.count
+            let repetitionCount = self.buffer.count - newBuffer.count
+            if repetitionCount > 0 {
+                newBuffer.append(contentsOf: repeatElement(nil, count: repetitionCount))
+            }
```

Comment:
This math is wrong. The following test fails, but should succeed:

```swift
    func testBufferManaged() {
        var ring = CircularBuffer<Int>(initialRingCapacity: 2)
        ring.append(1)
        XCTAssertEqual(ring.capacity, 2)

        // Now we want to replace the last subrange with two elements. This should
        // force an increase in size.
        ring.replaceSubrange(0..<1, with: [3, 4])
        XCTAssertEqual(ring.capacity, 4)
    }
```

The core issue here is that you are filling up to the *old* buffer size, but you need to fill up to the *new* one. `repetitionCount` should be calculated as `newCapacity - newBuffer.count`. That will also allow us to avoid a branch, as the number can never be negative: just append `repeatElement` unconditionally.

Reply:
Ugh!! shame cube.

Good catch Cory. I'll fix it during the day
Thanks